legacy/web/src/main/java/org/jboss/as/web/WebMigrateOperation.java

```diff
@@ -0,0 +1,725 @@
+        migrationOperations.clear();
+        for (Map.Entry<PathAddress, ModelNode> e : old.entrySet()) {
+            if (e.getKey().getElement(0).getKey().equals(SUBSYSTEM)) {
+                final PathAddress oldAddress = pathAddress(e.getValue().get(ADDRESS));
+                List<PathElement> elements = new ArrayList<>();
+                for (PathElement j : prefix) {
+                    elements.add(j);
+                }
+                for (PathElement j : oldAddress) {
+                    elements.add(j);
+                }
+                PathAddress newAddress = pathAddress(elements);
+                e.getValue().get(ADDRESS).set(newAddress.toModelNode());
+                migrationOperations.put(newAddress, e.getValue());
+            } else {
+                //not targeted at a subsystem
+                migrationOperations.put(e.getKey(), e.getValue());
+            }
+        }
+    }
+
+    /**
+     * We need to create the IO subsystem, if it does not already exist
+     */
+    private void createIoSubsystem(OperationContext context, Map<PathAddress, ModelNode> migrationOperations) {
+        Resource root = context.readResourceFromRoot(PathAddress.EMPTY_ADDRESS, false);
+        if (root.getChildrenNames(SUBSYSTEM).contains(IOExtension.SUBSYSTEM_NAME)) {
+            // subsystem is already added, do nothing
+            return;
+        }
+
+        PathAddress address = pathAddress(pathElement(SUBSYSTEM, IOExtension.SUBSYSTEM_NAME));
+        migrationOperations.put(address, createAddOp(address));
+        address = pathAddress(pathElement(SUBSYSTEM, IOExtension.SUBSYSTEM_NAME), pathElement("worker", "default"));
+        migrationOperations.put(address, createAddOp(address));
+        address = pathAddress(pathElement(SUBSYSTEM, IOExtension.SUBSYSTEM_NAME), pathElement("buffer-pool", "default"));
+        migrationOperations.put(address, createAddOp(address));
+
+    }
+
+
+    /**
+     * create a handler for serving welcome content
+     */
+    private void createWelcomeContentHandler(Map<PathAddress, ModelNode> migrationOperations) {
+
+        PathAddress address = pathAddress(pathElement(SUBSYSTEM, UndertowExtension.SUBSYSTEM_NAME), pathElement(Constants.CONFIGURATION, Constants.HANDLER));
+        migrationOperations.put(address, createAddOp(address));
+
+        address = pathAddress(pathElement(SUBSYSTEM, UndertowExtension.SUBSYSTEM_NAME), pathElement(Constants.CONFIGURATION, Constants.HANDLER), pathElement(Constants.FILE, "welcome-content"));
+        final ModelNode add = createAddOp(address);
+        add.get(Constants.PATH).set(new ModelNode(new ValueExpression("${jboss.home.dir}/welcome-content")));
+        migrationOperations.put(address, add);
+    }
+
+    private ModelNode createAddOp(PathAddress address) {
+        ModelNode add = new ModelNode();
+        add.get(OP).set(ADD);
+        add.get(ADDRESS).set(address.toModelNode());
+        return add;
+    }
```

Comment:
Not a problem, but there's org.jboss.as.controller.operations.common.Util.createAddOperation(PathAddress address)